lib/TikZImage.pm

```diff
@@ -166,49 +168,57 @@ sub draw {
 	close $fh;
 
 	my $ext = $self->ext;
-	my $tex_ext = $ext eq 'svg' && $self->svgMethod eq 'dvisvgm' ? 'dvi' : 'pdf';
-	my $latex_binary = WeBWorK::PG::IO::externalCommand($tex_ext eq 'dvi' ? 'latex' : 'pdflatex');
-
-	# Generate the pdf file.
-	system "cd " . $working_dir . " && $latex_binary image.tex > pdflatex.stdout 2> /dev/null";
-
-	if (-r "$working_dir/image.$tex_ext") {
-		chmod(0777, "$working_dir/image.$tex_ext");
-		# Convert the file to the appropriate type of image file
-		if ($ext eq 'svg') {
-			if ($self->svgMethod eq 'dvisvgm') {
-				system WeBWorK::PG::IO::externalCommand('dvisvgm') .
-					" $working_dir/image.dvi --no-fonts --output=$working_dir/image.svg > /dev/null 2>&1";
-			} else {
-				system WeBWorK::PG::IO::externalCommand($self->svgMethod) .
-					" $working_dir/image.pdf $working_dir/image.svg > /dev/null 2>&1";
-			}
-		} elsif ($ext ne 'pdf') {
-			system WeBWorK::PG::IO::externalCommand('convert') .
-				join('',map {" -$_ " . $self->convertOptions->{input}->{$_}} (keys %{$self->convertOptions->{input}})) .
-				" $working_dir/image.pdf" .
-				join('',map {" -$_ " . $self->convertOptions->{output}->{$_}} (keys %{$self->convertOptions->{output}})) .
-				" $working_dir/image.$ext > /dev/null 2>&1";
-		}
+	my $svgMethod = $self->svgMethod;
+	my $latex     = WeBWorK::PG::IO::externalCommand('latex');
+	my $pdflatex  = WeBWorK::PG::IO::externalCommand('pdflatex');
 
-		if (-r "$working_dir/image.$ext") {
-			# Read the generated image file into memory
-			open(my $in_fh,  "<", "$working_dir/image.$ext")
-				or warn "Failed to open $working_dir/image.$ext for reading.", return;
-			local $/;
-			$data = <$in_fh>;
-            close($in_fh);
-		} else {
-			warn "Convert operation failed.";
+	# Make only the dvi, only the pdf, or both in case we are making tgz with svg via dvisvgm
+	if (($ext eq 'svg' || $ext eq 'tgz') && $svgMethod eq 'dvisvgm') {
+		system "cd $working_dir && $latex image.tex > latex.stdout 2> /dev/null";
+	}
+	if ($ext ne 'svg' || ($ext eq 'svg' && $svgMethod ne 'dvisvgm')) {
+		system "cd $working_dir && $pdflatex image.tex > pdflatex.stdout 2> /dev/null";
+	}
+
+	# Make derivatives of the dvi
+	if (-r "$working_dir/image.dvi") {
+		if (($ext eq 'svg' || $ext eq 'tgz') && $svgMethod eq 'dvisvgm') {
+			$self->use_svgMethod($working_dir, "image.dvi");
 		}
 	} else {
-		warn "File $working_dir/image.$tex_ext was not created.";
-		if (open(my $err_fh, "<", "$working_dir/pdflatex.stdout")) {
-			while (my $error = <$err_fh>) {
-				warn $error;
-			}
-			close($err_fh);
+		warnLaTeXOutput($working_dir, "image.dvi");
+	}
```

Comment:
The logic is not correct here.  This will add warnings to the output anytime that the dvi file does not exist, even when that file should not exist.  For example, if a png image is requested or if an svg image is requested via pdf2svg.  The inner `if` here should be the outer `if`, and the `else` should belong to the inner `if`.  I.e.,

```suggestion
if (($ext eq 'svg' || $ext eq 'tgz') && $svgMethod eq 'dvisvgm') {
	if (-r "$working_dir/image.dvi") {
			$self->use_svgMethod($working_dir, "image.dvi");
	} else {
		warnLaTeXOutput($working_dir, "image.dvi");
	}
}
```